File: Programmers154540.java
import java.util.*;
class Point {
    int x;
    int y;
    Point(int x, int y){
        this.x = x;
        this.y = y;
    }
}

class Solution {
    static int[] emptyIsland = {-1};
    public int[] solution(String[] maps) {
        List<Integer> days = new ArrayList<>();
        char[][] islandMap = new char[maps.length][maps[0].length()];
        boolean[][] visited = new boolean[maps.length][maps[0].length()];
        
        for(int i=0;i<islandMap.length;i++){
            for(int j=0;j<islandMap[i].length;j++){
                islandMap[i][j] = maps[i].charAt(j);
            }
        }
        int[] nx = {1,-1,0,0};
        int[] ny = {0,0,1,-1};
        
        for(int i=0;i<islandMap.length;i++){
            for(int j=0;j<islandMap[i].length;j++){
                if(Character.isDigit(islandMap[i][j]) && !visited[i][j]){
                    Queue<Point> q = new LinkedList<>();
                    q.add(new Point(i,j));
                    int sum = 0;
                    
                    while(!q.isEmpty()){
                        Point c = q.poll();
                        if (visited[c.x][c.y]) continue; // 이미 방문했으면 건너뛰기
                        visited[c.x][c.y] = true; // 현재 위치를 방문한 것으로 표시
                        int currentDay = islandMap[c.x][c.y] - '0';
                        sum += currentDay; // 현재 지점을 sum에 추가

                        for(int move=0;move<4;move++){
                            int mx = c.x + nx[move];
                            int my = c.y + ny[move];
                            // 올바른 범위 검사 조건
                            if(mx >= 0 && my >= 0 && mx < islandMap.length && my < islandMap[0].length){
                                if(Character.isDigit(islandMap[mx][my]) && !visited[mx][my]){
                                    q.add(new Point(mx,my));
                                }
                            }
                        }
                    }
                    days.add(sum);
                }
            }
        }

        if(days.size() == 0) return emptyIsland;
        
        Collections.sort(days);
        int[] answer = new int[days.size()];
        for(int i=0;i<days.size();i++){
            answer[i] = days.get(i);
        }
        
        return answer;
    }
}

File: Programmers154540.js
function solution(maps) {
    var rowLength = maps.length;
    var colLength = maps[0].length;
    const chMap = Array.from(Array(rowLength), 
                       () => Array(colLength).fill(null));
    //방문 체크
    const visited = Array.from(Array(rowLength), 
                       () => Array(colLength).fill(false));
    
    for(let i = 0; i < rowLength; i++){
        for(let j = 0; j < colLength; j++){
            chMap[i][j] = maps[i][j];
        }
    }

    //동서남북
    const nx = [1, -1, 0, 0];
    const ny = [0, 0, 1, -1];
    const days = [];
    
    for(let i = 0; i < rowLength; i++){
        for(let j = 0; j < colLength; j++){
            if(!isNaN(chMap[i][j]) && !visited[i][j]){
                const q = [{x : i, y : j}];
                let sum = 0;
                while(q.length > 0){
                    const {x, y} = q.shift();
                    if(visited[x][y]) continue;
                    visited[x][y] = true;
                    
                    const currentDay = parseInt(chMap[x][y],10);
                    sum += currentDay;
                    for(let move =0;move<4;move++){
                        const mx = x + nx[move];
                        const my = y + ny[move];
                        if(mx >= 0 && mx < maps.length && 
                           my >= 0 && my < maps[0].length
                          ){
                            if(!isNaN(chMap[mx][my]) && !visited[mx][my]){
                                q.push({ x : mx, y : my});
                                //visited[mx][my] = true;
                            }
                        }
                    }
                }
                days.push(sum);
            }
        }
    }
    if(days.length === 0) return [-1];
    
    days.sort((a, b) => a - b);
    return days;
}

